lib/request.js

```diff
@@ -7,67 +7,57 @@
 const InvalidArgumentError = require('./errors/invalid-argument-error');
 const typeis = require('type-is');
 
-/**
- * Constructor.
- */
+class Request {
+  constructor({ headers, method, query, body, ...otherOptions } = {}) {
+    if (!headers) {
+      throw new InvalidArgumentError('Missing parameter: `headers`');
+    }
 
-function Request(options) {
-  options = options || {};
+    if (!method) {
+      throw new InvalidArgumentError('Missing parameter: `method`');
+    }
 
-  if (!options.headers) {
-    throw new InvalidArgumentError('Missing parameter: `headers`');
-  }
+    if (!query) {
+      throw new InvalidArgumentError('Missing parameter: `query`');
+    }
 
-  if (!options.method) {
-    throw new InvalidArgumentError('Missing parameter: `method`');
+    this.body = body || {};
+    this.headers = {};
+    this.method = method;
+    this.query = query;
+
+    // Store the headers in lower case.
+    Object.entries(headers).forEach(([header, value]) => {
+      this.headers[header.toLowerCase()] = value;
+    });
+
+    // Store additional properties of the request object passed in
+    Object.entries(otherOptions)
+      .filter(([property]) => !this[property])
```

Comment:
We need to use `Object.prototype.hasOwnProperty.call(options, property)` anyway here, to avoid manipulation of the prototype chain, especially because Request options may be built from a client.

Reply:
@jankapunkt - `Object.entries()` (like `Object.keys()`) does not traverse the prototype chain but only iterates over the own enumerable properties of the object.

[MDN docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/entries#description):
> `Object.entries()` returns an array whose elements are arrays corresponding to the enumerable string-keyed property key-value pairs found directly upon object. This is the same as iterating with a [for...in](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in) loop, except that a for...in loop enumerates properties in the prototype chain as well. The order of the array returned by Object.entries() is the same as that provided by a [for...in](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in) loop.

In other words, I believe the `hasOwnProperty` check will always return true for anything we iterate other with `Object.entries()`, and therefore it is redundant/not necessary. It is certainly possible I'm misunderstanding the concern you have in mind, though, so please let me know if I'm missing the mark!

Reply:
@jorenvandeweyer - You're right that it's not strictly necessary to check again; I had the same thought while I was refactoring.

The reason I ultimately left the existing check in place is because I worried that trusting the destructuring as a safeguard is a bit brittle -- I could easily see someone in the future changing which options are destructured without realizing that it could have a side effect on the "custom property" logic.

I'm happy to defer to the maintainers' preference, though. So, let me know if you would prefer for me to remove the check, and I'll do it.

Reply:
@menewman you are correct but take a close look which own properties are used by Object.entries - the ones from the options but the assignment goes to the request instance.

Therefore I could tamper properties from the chain like this:

```js
const req = new Request({ body, headers, toString: () => '[object IAMValid]' })
req.toString() // [object IAMValid]
```

This is because `toString` is an own property of the options but for request it's part of the prototype chain. This could theoretically allow to tamper the chain.

I know this is very superficial but I think we are on the safe side, adding the extra check

Reply:
@jorenvandeweyer I will add this to the next PR for the improved tests (#219)

Reply:
@jankapunkt - I don't think that's possible because of the `.filter(([property]) => !this[property])` check.

With the combo of Object.entries + this filter function...
- We only loop over enumerable keys on options
- We only overwrite keys on `this` that do not already exist (including directly on the object, or on the prototype chain)

I believe this is functionally identical to the prior version of the logic with the for...in + hasOwnProperty check.

But the best proof is always in tests, so I also added a couple of unit tests to check this case specifically in my latest commit!

Reply:
https://github.com/node-oauth/node-oauth2-server/pull/225/commits/8ea66994831892dc28ec0215246ecd2a3752f08c

Reply:
@menewman thank you I understand it now and thanks for the added tests